pkg/ebpf/generic_map.go

```diff
@@ -0,0 +1,334 @@
+		itops.BatchSize = int(g.m.MaxEntries())
+	}
+
+	if BatchAPISupported() && !g.isPerCPU() && !itops.ForceSingleItem {
+		it := &genericMapBatchIterator[K, V]{
+			m:                            g.m,
+			batchSize:                    itops.BatchSize,
+			keys:                         make([]K, itops.BatchSize),
+			values:                       make([]V, itops.BatchSize),
+			valueTypeCanUseUnsafePointer: g.valueTypeCanUseUnsafePointer(),
+		}
+
+		// Do an initial copy of the keys/values slices to avoid allocations
+		it.keysCopy = it.keys
+		it.valuesCopy = it.values
+
+		return it
+	}
+
+	return &genericMapItemIterator[K, V]{
+		it:                           g.m.Iterate(),
+		valueTypeCanUseUnsafePointer: g.valueTypeCanUseUnsafePointer(),
+	}
+}
+
+// IteratePerCPU returns an iterator for the map (only to be used in per-cpu maps)
+func (g *GenericMap[K, V]) IteratePerCPU() GenericMapIterator[K, []V] {
+	return &genericMapItemIterator[K, []V]{
+		it:                           g.m.Iterate(),
+		valueTypeCanUseUnsafePointer: false,
+	}
+}
+
+type genericMapItemIterator[K interface{}, V interface{}] struct {
+	it                           *ebpf.MapIterator
+	valueTypeCanUseUnsafePointer bool
+}
+
+func (g *genericMapItemIterator[K, V]) Next(key *K, value *V) bool {
+	// we resort to unsafe.Pointers because by doing so the underlying eBPF
+	// library avoids marshaling the key/value variables while traversing the map
+	// However, in some cases (slices, structs) we need to pass the variable directly
+	// so that the library detects the type correctly
+	if g.valueTypeCanUseUnsafePointer {
+		return g.it.Next(unsafe.Pointer(key), unsafe.Pointer(value))
+	}
+
+	return g.it.Next(unsafe.Pointer(key), value)
+}
+
+func (g *genericMapItemIterator[K, V]) Err() error {
+	return g.it.Err()
+}
+
+type genericMapBatchIterator[K interface{}, V interface{}] struct {
+	m                            *ebpf.Map
+	batchSize                    int
+	cursor                       ebpf.BatchCursor
+	keys                         []K
+	values                       []V
+	keysCopy                     any // A pointer to keys of type "any", used to avoid allocations when calling BatchLookup
+	valuesCopy                   any
+	currentBatchSize             int
+	inBatchIndex                 int
+	err                          error
+	totalCount                   int
+	lastBatch                    bool
+	valueTypeCanUseUnsafePointer bool
+}
+
+func (g *genericMapBatchIterator[K, V]) Next(key *K, value *V) bool {
+	// Safety check to avoid an infinite loop
+	if g.totalCount >= int(g.m.MaxEntries()) {
+		return false
+	}
+
+	// We have finished all the values in the current batch (or there wasn't any batch
+	// to begin with, with g.currentBatchSize == 0), so we need to fetch the next batch
+	if g.inBatchIndex >= g.currentBatchSize {
+		if g.lastBatch {
+			return false
+		}
+
+		// Important! If we pass here g.keys/g.values, Go will create a copy of the slice
```

Comment:
you mean a new slice instance will be created, not the underlying buffer of the slice. since slice just holds the pointer to the underlying buffer

Reply:
Yes, updating the comment.